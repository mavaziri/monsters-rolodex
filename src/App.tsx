// import { Component } from 'react';
import { useState, useEffect, ChangeEvent } from 'react';

// import axios from 'axios';
import { useQuery } from '@tanstack/react-query';

import CardList from './components/card-list/card-list.component';
import SearchBox from './components/search-box/search-box.component';

import { getData } from './utils/data.utils';
import './App.css';

export type Monster = {
  id: string;
  name: string;
  email: string;
};

const App = () => {
  const [searchField, setSearchField] = useState('');
  const [filteredMonsters, setFilteredMonsters] = useState<Monster[]>([]);

  const {
    data: monsters,
    isLoading,
    error,
  } = useQuery<Monster[]>({
    queryKey: ['users'],
    queryFn: async (): Promise<Monster[]> => {
      return await getData<Monster[]>(
        'https://jsonplaceholder.typicode.com/users'
      );
    },
  });

  if (error) {
    console.error('Error fetching data:', error.message);
  }

  useEffect(() => {
    if (monsters) {
      const newFilteredMonsters = monsters.filter((monster: Monster) =>
        monster.name.toLowerCase().includes(searchField.toLowerCase())
      );

      setFilteredMonsters(newFilteredMonsters);
    }
  }, [monsters, searchField]);

  const onSearchChange = (event: ChangeEvent<HTMLInputElement>): void => {
    const searchFieldString = event.target.value.toLocaleLowerCase();
    setSearchField(searchFieldString);
  };

  return isLoading ? (
    <h2>Loading...</h2>
  ) : error ? (
    <h2>Error loading monsters</h2>
  ) : (
    <div className="App">
      <SearchBox
        className="monsters-search-box"
        onChangeHandler={onSearchChange}
        placeholder="search monsters"
      />
      <CardList monsters={filteredMonsters} />
    </div>
  );
};

export default App;

// class App extends Component {
//   constructor() {
//     super();

//     this.state = {
//       monsters: [],
//       searchField: '',
//     };

//     // console.log('constructor');
//   }

//   componentDidMount() {
//     // fetch('https://jsonplaceholder.typicode.com/users').then((response) =>
//     //   response.json().then((users) =>
//     //     this.setState(
//     //       () => {
//     //         return { monsters: users };
//     //       },
//     //       () => {
//     //         console.log(this.state);
//     //       }
//     //     )
//     //   )
//     // );
//     // console.log('componentDidMount');

//     axios.get('https://jsonplaceholder.typicode.com/users').then((response) => {
//       this.setState(() => {
//         return { monsters: response.data };
//       });
//     });
//   }

//   onSearchChange = (event) => {
//     this.setState({ searchField: event.target.value });
//   };

//   render() {
//     // console.log('render');
//     const { monsters, searchField } = this.state;
//     const { onSearchChange } = this;

//     const filteredMonsters = monsters.filter((monster) =>
//       monster.name.toLowerCase().includes(searchField.toLowerCase())
//     );

//     return (
//       <div className="App">
//         <h1 className="app-title">Monsters Rolodex</h1>

//         <SearchBox
//           className="monsters-search-box"
//           onChangeHandler={onSearchChange}
//           placeholder="search monsters"
//         />
//         <CardList monstersProp={filteredMonsters} />
//       </div>
//     );
//   }
// }
