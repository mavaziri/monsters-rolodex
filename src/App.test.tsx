import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { useQuery } from '@tanstack/react-query';
import App, { Monster } from './App';
import { error } from 'console';

const mockMonsters: Monster[] = [
  { id: '1', name: 'Monster One', email: 'monster1@example.com' },
  { id: '2', name: 'Monster Two', email: 'monster2@example.com' },
  { id: '3', name: 'Monster Three', email: 'monster3@example.com' },
];

jest.mock('@tanstack/react-query', () => ({
  ...jest.requireActual('@tanstack/react-query'),
  useQuery: jest.fn(),
}));

// Create a function to return a new query client for each test
const createTestQueryClient = () =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false, // Disable retries to simplify testing
      },
    },
  });

export const renderWithQueryClient = (ui: React.ReactElement) => {
  const queryClient = createTestQueryClient();
  return render(
    <QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>
  );
};

describe('App Component', () => {
  // beforeEach(() => {
  //   jest.clearAllMocks(); // Reset mocks before each test
  // });
  beforeEach(() => {
    (useQuery as jest.Mock).mockReturnValue({
      data: undefined, // No monsters initially
      isLoading: false, // Not in loading state
      error: null, // No error
    });
  });

  test('expect to render App component', () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  test('displays loading state when data is being fetched', () => {
    (useQuery as jest.Mock).mockReturnValue({
      data: undefined,
      isLoading: true,
      error: null,
    });

    renderWithQueryClient(<App />);

    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  test('displays error message when data fetching fails', () => {
    (useQuery as jest.Mock).mockReturnValue({
      data: undefined,
      isLoading: false,
      error: new Error('Network error'), // Simulate error
    });

    renderWithQueryClient(<App />);

    expect(screen.getByText('Error loading monsters')).toBeInTheDocument();
  });

  test('displays the list of monsters after data is loaded', async () => {
    (useQuery as jest.Mock).mockReturnValue({
      data: mockMonsters,
      isLoading: false,
      error: null,
    });

    renderWithQueryClient(<App />);

    await waitFor(() => {
      expect(screen.getByText('Monster One')).toBeInTheDocument();
      expect(screen.getByText('Monster Two')).toBeInTheDocument();
      expect(screen.getByText('Monster Three')).toBeInTheDocument();
    });
  });

  test('displays filtered monsters based on search fields', async () => {
    (useQuery as jest.Mock).mockReturnValue({
      data: mockMonsters,
      isLoading: false,
      error: null,
    });

    renderWithQueryClient(<App />);

    // Simulate typing "One" in the search box
    const searchBox = screen.getByPlaceholderText('search monsters');
    fireEvent.change(searchBox, { target: { value: 'One' } });

    // Wait for filtered results
    await waitFor(() => {
      expect(screen.getByText('Monster One')).toBeInTheDocument();
      expect(screen.queryByText('Monster Two')).not.toBeInTheDocument();
      expect(screen.queryByText('Monster Three')).not.toBeInTheDocument();
      //or
      // expect(screen.queryByText('Monster Two')).toBeNull();
      // expect(screen.queryByText('Monster Three')).toBeNull();
    });
  });

  test('displays no monsters when search query has no matches', async () => {
    (useQuery as jest.Mock).mockReturnValue({
      data: mockMonsters,
      isLoading: false,
      error: null,
    });

    renderWithQueryClient(<App />);

    // Simulate typing a query that doesn't match any monster
    const searchBox = screen.getByPlaceholderText('search monsters');
    fireEvent.change(searchBox, { target: { value: 'NonExistent' } });

    await waitFor(() => {
      // Assert no monsters are shown
      expect(screen.queryByText('Monster One')).not.toBeInTheDocument();
      expect(screen.queryByText('Monster Two')).not.toBeInTheDocument();
      expect(screen.queryByText('Monster Three')).not.toBeInTheDocument();
    });
  });

  // test('displays retry button on error and reloads data when clicked', async () => {
  //   (useQuery as jest.Mock).mockReturnValue({
  //     data: undefined,
  //     isLoading: false,
  //     error: null,
  //   });

  //   renderWithQueryClient(<App />);

  //   // Verify error message is displayed
  //   expect(screen.getByText('Error loading monsters')).toBeInTheDocument();

  //   // Simulate clicking retry (assuming a retry button exists)
  //   fireEvent.click(screen.getByText('Retry'));

  //   // Mock successful data loading
  //   (useQuery as jest.Mock).mockReturnValue({
  //     data: mockMonsters,
  //     isLoading: false,
  //     error: null,
  //   });

  //   await waitFor(() => {
  //     expect(screen.getByText('Monster One')).toBeInTheDocument();
  //     expect(screen.getByText('Monster Two')).toBeInTheDocument();
  //     expect(screen.getByText('Monster Three')).toBeInTheDocument();
  //   });
  // });

  test('does not display search box or monsters while loading', () => {
    // Mock the loading state for this specific test
    (useQuery as jest.Mock).mockReturnValue({
      data: undefined,
      isLoading: true,
      error: null,
    });

    renderWithQueryClient(<App />);

    // Assert that the loading message is displayed
    expect(screen.getByText('Loading...')).toBeInTheDocument();

    // Assert that the search box and monster list are not rendered
    expect(
      screen.queryByPlaceholderText('search monsters')
    ).not.toBeInTheDocument();
    expect(screen.queryByText('Monster One')).not.toBeInTheDocument();
    expect(screen.queryByText('Monster Two')).not.toBeInTheDocument();
    expect(screen.queryByText('Monster Three')).not.toBeInTheDocument();
  });

  test('initial render has empty search field and no monsters displayed', () => {
    (useQuery as jest.Mock).mockReturnValue({
      data: undefined,
      isLoading: false,
      error: null,
    });

    renderWithQueryClient(<App />);

    // Assert that the search box is present and empty
    const searchBox = screen.getByPlaceholderText('search monsters');
    expect(searchBox).toBeInTheDocument();
    expect(searchBox).toHaveValue('');

    // Assert that no monsters are displayed
    expect(screen.queryByText('Monster One')).not.toBeInTheDocument();
    expect(screen.queryByText('Monster Two')).not.toBeInTheDocument();
  });
});
