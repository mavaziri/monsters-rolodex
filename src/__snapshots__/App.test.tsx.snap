// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component expect to render App component 1`] = `
<div>
  <div
    class="App"
  >
    <input
      class="search-box monsters-search-box"
      placeholder="search monsters"
      type="search"
    />
    <div
      class="card-list"
    />
  </div>
</div>
`;
